File: model/src/main/kotlin/Environment.kt
/*
 * Copyright (C) 2017-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.ort.model

import com.here.ort.utils.OS

/**
 * A description of the environment that ORT was executed in.
 */
data class Environment(
        /**
         * The version of ORT used.
         */
        val ortVersion: String = ORT_VERSION,

        /**
         * Name of the operating system, defaults to [OS.name].
          */
        val os: String = OS.name,

        /**
         * Map of selected environment variables that might be relevant for debugging.
         */
        val variables: Map<String, String> = System.getenv().mapKeys { (key, _) -> key.toUpperCase() }.let { env ->
            RELEVANT_VARIABLES.mapNotNull { key ->
                env[key]?.let { value -> key to value }
            }.toMap()
        },

        /**
         * Map of used tools and their installed versions, defaults to an empty map.
         */
        val toolVersions: Map<String, String> = emptyMap()
) {
    companion object {
        val ORT_VERSION = this::class.java.getResource("/VERSION").readText()

        private val RELEVANT_VARIABLES = listOf(
                // Windows variables.
                "OS",
                "COMSPEC",
                // Unix variables.
                "OSTYPE",
                "HOSTTYPE",
                "SHELL",
                "TERM",
                // General variables.
                "JAVA_HOME",
                "ANDROID_HOME",
                "GOPATH"
        )
    }
}

File: reporter-web-app/src/models/Environment.js
/*
 * Copyright (C) 2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

class Environment {
    #ortVersion = '';

    #javaVersion = '';

    #os = '';

    #variables = new Map();

    constructor(obj) {
        if (obj instanceof Object) {
            if (obj.ort_version) {
                this.ortVersion = obj.ort_version;
            }

            if (obj.ortVersion) {
                this.ortVersion = obj.ortVersion;
            }

            if (obj.java_version) {
                this.javaVersion = obj.java_version;
            }

            if (obj.javaVersion) {
                this.javaVersion = obj.javaVersion;
            }

            if (obj.os) {
                this.os = obj.os;
            }

            if (obj.variables) {
                this.variables = obj.variables;
            }
        }
    }

    get ortVersion() {
        return this.#ortVersion;
    }

    set ortVersion(val) {
        this.#ortVersion = val;
    }

    get javaVersion() {
        return this.#javaVersion;
    }

    set javaVersion(val) {
        this.#javaVersion = val;
    }

    get os() {
        return this.#os;
    }

    set os(val) {
        this.#os = val;
    }

    get variables() {
        return this.#variables;
    }

    set variables(val) {
        this.#variables = new Map(Object.entries(val));
    }
}

export default Environment;
